### ISSUE TITLE
Can't save setting in FF67

### ISSUE THREAD

**Issue by ooker777** [hnakamur/FormatLink-Firefox#25]
I try this:

![](https://i.imgur.com/t9lqHWX.png)

with and without checking `Create submenus`, but it resets after saved. The panel is blank:

![](https://i.imgur.com/9NkQzwS.png)

**Comment by scruss** [comment]
Same for me, FF 66.0.3 on Ubuntu 19.04 x86_64

**Comment by scivola** [comment]
Same for me, FF 66.0.3 on macOS 10.13.6.

**Comment by scivola** [comment]
Now it works.
(from Firefox 66.0.4 ?)

**Comment by scivola** [comment]
It does not work, again. (same as @ooker777)
Firefox 66.0.5, macOS 10.13.6

**Comment by scruss** [comment]
broken again

FF 67.0 (Ubuntu), Format Link 3.0.1

**Comment by akabekobeko** [comment]
It seems to occur when create a new environment. I commented that on #20.

**Comment by mavensecurity** [comment]
Ditto -It does not let you save settings; there are no default settings; It does nothing at this point. Format Link v3.0.1 in FF 64.0.4 Windows

**Comment by scruss** [comment]
Oddly, I have one Firefox instance where FormatLink does still work perfectly: Firefox 67.0.4, Windows 10, FormatLink v3.0.1 (May 7, 2019 - automatic updates set to **Default** rather than **On**). I wish I knew where the difference was between this and my other installations that don't work.

**Comment by lo48576** [comment]
Reproduced in Firefox 70.0b14 (64-bit).

I checked stderr log of `/usr/bin/firefox`, and found that the lines below are logged when I pushed "Save changes" button.

```
console.warn: LoginRecipes: "getRecipes: falling back to a synchronous message for:" "moz-extension://feae02b2-7613-4beb-bc87-40e07046d63c"
WARNING: content window passed to PrivateBrowsingUtils.isWindowPrivate. Use isContentWindowPrivate instead (but only for frame scripts).
pbu_isWindowPrivate@resource://gre/modules/PrivateBrowsingUtils.jsm:27:11
isBrowserPrivate@resource://gre/modules/PrivateBrowsingUtils.jsm:47:19
fire@resource://gre/modules/WebNavigation.jsm:553:30
onDocumentChange@resource://gre/modules/WebNavigation.jsm:508:10
receiveMessage@resource://gre/modules/WebNavigation.jsm:377:14
WARNING: content window passed to PrivateBrowsingUtils.isWindowPrivate. Use isContentWindowPrivate instead (but only for frame scripts).
pbu_isWindowPrivate@resource://gre/modules/PrivateBrowsingUtils.jsm:27:11
isBrowserPrivate@resource://gre/modules/PrivateBrowsingUtils.jsm:47:19
fire@resource://gre/modules/WebNavigation.jsm:553:30
onDocumentChange@resource://gre/modules/WebNavigation.jsm:508:10
receiveMessage@resource://gre/modules/WebNavigation.jsm:377:14
WARNING: content window passed to PrivateBrowsingUtils.isWindowPrivate. Use isContentWindowPrivate instead (but only for frame scripts).
pbu_isWindowPrivate@resource://gre/modules/PrivateBrowsingUtils.jsm:27:11
isBrowserPrivate@resource://gre/modules/PrivateBrowsingUtils.jsm:47:19
fire@resource://gre/modules/WebNavigation.jsm:553:30
onLoad@resource://gre/modules/WebNavigation.jsm:527:10
receiveMessage@resource://gre/modules/WebNavigation.jsm:385:14
JavaScript error: moz-extension://feae02b2-7613-4beb-bc87-40e07046d63c/common.js, line 29: TypeError: invalid assignment to const `options'
```

It seems that the line 27 of `common.js` should use `let options = ...` instead of `const options = ...` (because it could be rewritten at line 29).

https://github.com/hnakamur/FormatLink-Firefox/blob/3291a7c0715e6a0a50a37433222b350759731940/common.js#L27-L29

**Comment by lo48576** [comment]
# Quick workaround

1. Open `about:debugging#/runtime/this-firefox`.
    + type this in location bar and press Enter key.
2. Push "Inspect" button for "Format Link" extension.
    + It exists in "Extensions" section.
    + This opens developer console, with extension scripts loaded.
3. Select "Console" tab of developer console, type `await browser.storage.sync.set(DEFAULT_OPTIONS);`, and press Enter key.
    + `DEFAULT_OPTIONS` is defined at line 3 of `common.js`.
      https://github.com/hnakamur/FormatLink-Firefox/blob/3291a7c0715e6a0a50a37433222b350759731940/common.js#L3-L24
    + This statement forcibly saves the default settings to internal storage for "Format Link" extension.
    + This makes `Object.keys(options).length` non-zero at line 28 of `common.js`. Line 29 won't be executed and you will have no error.
      https://github.com/hnakamur/FormatLink-Firefox/blob/3291a7c0715e6a0a50a37433222b350759731940/common.js#L28
4. Now you will see default settings are saved in `about:addons`.
   You can modify and save custom settings.

* * *

By this workaround, the extension started to work as expected in my environment.

**Comment by nopoz** [comment]
> 
> # Quick workaround
> 
>     1. Open `about:debugging#/runtime/this-firefox`.
>        
>        * type this in location bar and press Enter key.
> 
>     2. Push "Inspect" button for "Format Link" extension.
>        
>        * It exists in "Extensions" section.
>        * This opens developer console, with extension scripts loaded.
> 
>     3. Select "Console" tab of developer console, type `await browser.storage.sync.set(DEFAULT_OPTIONS);`, and press Enter key.
>        
>        * `DEFAULT_OPTIONS` is defined at line 3 of `common.js`.
>          https://github.com/hnakamur/FormatLink-Firefox/blob/3291a7c0715e6a0a50a37433222b350759731940/common.js#L3-L24
>        * This statement forcibly saves the default settings to internal storage for "Format Link" extension.
>        * This makes `Object.keys(options).length` non-zero at line 28 of `common.js`. Line 29 won't be executed and you will have no error.
>          https://github.com/hnakamur/FormatLink-Firefox/blob/3291a7c0715e6a0a50a37433222b350759731940/common.js#L28
> 
>     4. Now you will see default settings are saved in `about:addons`.
>        You can modify and save custom settings.
> 
> 
> By this workaround, the extension started to work as expected in my environment.

I don't understand this workaround - what am I supposed to be changing? Using FF74, I can't save any settings.

**Comment by lo48576** [comment]
@nopoz You are supposed to change (initialize) the internal storage of FormatLink addon.

Currently FormatLink has a bug and it cannot initialize its internal storage through plugin's usual setting page.
As a workaround for this, you can initialize the storage manually not by plugin's setting page, but by directly running a program to do that.

The workaround above is a procedure for doing manual initialization of the storage.

**Comment by hnakamur** [comment]
Sorry for being late, I added the fix as commented at https://github.com/hnakamur/FormatLink-Firefox/issues/25#issuecomment-541430088 and published version 4.0.2.
Could you upgrade to the latest version?
Thanks!

**Comment by hnakamur** [comment]
Sorry for not responding. I'm closing this since it's too old.
